File: sdk/python/dghub_sdk/codec.py
"""消息数据类与编解码器（序列化 / 反序列化）。"""

import json
from dataclasses import dataclass
from enum import StrEnum
from typing import Any

from .enums import Action, Channel, DeviceType, LogLevel, OpCode, StrengthMode


# ---------------------------------------------------------------------------
# CodecMessage
# ---------------------------------------------------------------------------


@dataclass(slots=True)
class CodecMessage:
    """``Codec.parse()`` 的类型化结果。通过 ``.op`` 判断哪些字段有值。"""
    op: OpCode
    # hello_ack
    status: str | None = None
    # config
    data: dict[str, Any] | None = None
    # config_changed
    key: str | None = None
    value: bool | int | str | None = None
    # device_info
    connected: bool | None = None
    device_type: DeviceType | None = None
    max_strength_a: int | None = None
    max_strength_b: int | None = None
    # stop
    reason: str | None = None
    # ping
    t: float | None = None


# ---------------------------------------------------------------------------
# Codec (namespace class)
# ---------------------------------------------------------------------------


class Codec:
    """所有消息编解码的命名空间类，无需实例化。"""

    @staticmethod
    def hello(token: str, manifest: dict[str, Any]) -> str:
        """构建 hello 握手消息。"""
        if not token:
            raise ValueError("token is required")
        if not isinstance(manifest, dict):
            raise TypeError("manifest must be a dict")
        return json.dumps({
            "op": "hello",
            "token": token,
            "manifest": manifest,
        })

    @staticmethod
    def trigger(
        action: Action = Action.BOTH,
        delta_pct: int = 0,
        strength_mode: StrengthMode = StrengthMode.ROLLBACK,
        duration_s: float = 1.0,
        preset: str = "",
        channel: Channel = Channel.BOTH,
        label: str | None = None,
        username: str | None = None,
        name: str | None = None,
        cause: str | None = None,
        pulse_name: str | None = None,
        target_id: str | None = None,
    ) -> str:
        """构建统一触发消息。

        当 ``action`` 包含波形（Action.BOTH 或 Action.WAVEFORM）且
        ``preset`` 为空时抛出 ``ValueError``。
        """
        if action in (Action.BOTH, Action.WAVEFORM) and not preset:
            raise ValueError("preset is required when action includes waveform")
        msg: dict[str, Any] = {
            "op": "trigger",
            "action": action.value,
            "delta_pct": delta_pct,
            "strength_mode": strength_mode.value,
            "duration_s": duration_s,
            "preset": preset,
            "channel": channel.value,
        }
        for key, value in (
            ("label", label),
            ("username", username),
            ("name", name),
            ("cause", cause),
            ("pulse_name", pulse_name),
            ("target_id", target_id),
        ):
            if value is not None:
                msg[key] = value
        return json.dumps(msg)

    @staticmethod
    def event(
        label: str,
        name: str,
        username: str | None = None,
        strength_pct: int | None = None,
        duration: float = 1.0,
        event_id: str | None = None,
        cause: str | None = None,
        pulse_name: str | None = None,
        from_pct: int | None = None,
        to_pct: int | None = None,
        delta_pct: int | None = None,
        target_id: str | None = None,
    ) -> str:
        """构建一次性事件消息。"""
        if not label:
            raise ValueError("label is required")
        if not name:
            raise ValueError("name is required")
        msg: dict[str, Any] = {
            "op": "event",
            "label": label,
            "name": name,
            "duration": duration,
        }
        for key, value in (
            ("username", username),
            ("strength_pct", strength_pct),
            ("event_id", event_id),
            ("cause", cause),
            ("pulse_name", pulse_name),
            ("from_pct", from_pct),
            ("to_pct", to_pct),
            ("delta_pct", delta_pct),
            ("target_id", target_id),
        ):
            if value is not None:
                msg[key] = value
        return json.dumps(msg)

    @staticmethod
    def pulse(
        preset: str,
        channel: Channel = Channel.BOTH,
        target_id: str | None = None,
    ) -> str:
        """构建仅波形的脉冲消息。"""
        if not preset:
            raise ValueError("preset is required")
        msg = {
            "op": "pulse",
            "preset": preset,
            "channel": channel.value,
        }
        if target_id is not None:
            msg["target_id"] = target_id
        return json.dumps(msg)

    @staticmethod
    def set_strength(
        channel: Channel,
        pct: int,
        target_id: str | None = None,
    ) -> str:
        """构建 set_strength 消息。"""
        if not 0 <= pct <= 100:
            raise ValueError("pct must be 0-100")
        msg = {
            "op": "set_strength",
            "channel": channel.value,
            "pct": pct,
        }
        if target_id is not None:
            msg["target_id"] = target_id
        return json.dumps(msg)

    @staticmethod
    def adjust_strength(
        channel: Channel,
        delta_pct: int,
        target_id: str | None = None,
    ) -> str:
        """构建 adjust_strength 消息。"""
        if not -100 <= delta_pct <= 100:
            raise ValueError("delta_pct must be -100 to 100")
        msg = {
            "op": "adjust_strength",
            "channel": channel.value,
            "delta_pct": delta_pct,
        }
        if target_id is not None:
            msg["target_id"] = target_id
        return json.dumps(msg)

    @staticmethod
    def status(fields: dict[str, Any]) -> str:
        """构建状态上报消息。"""
        if not isinstance(fields, dict):
            raise TypeError("fields must be a dict")
        return json.dumps({
            "op": "status",
            "fields": fields,
        })

    @staticmethod
    def log(level: LogLevel, message: str) -> str:
        """构建日志消息。"""
        return json.dumps({
            "op": "log",
            "level": level.value,
            "message": message,
        })

    @staticmethod
    def set_config(key: str, value: Any) -> str:
        """构建 set_config 消息，用于持久化运行时数据。"""
        return json.dumps({
            "op": "set_config",
            "key": key,
            "value": value,
        })

    @staticmethod
    def parse(raw: str) -> CodecMessage:
        """将 JSON 字符串解析为类型化的 ``CodecMessage`` 数据类。"""
        data = json.loads(raw)
        match data:
            case {"op": "hello_ack"} as ack:
                ack_fields = {k: v for k, v in ack.items() if k != "op"}
                return CodecMessage(op=OpCode.HELLO_ACK, data=ack_fields)
            case {"op": "config", "data": d}:
                return CodecMessage(op=OpCode.CONFIG, data=d)
            case {"op": "config_changed", "key": k, "value": v}:
                return CodecMessage(op=OpCode.CONFIG_CHANGED, key=k, value=v)
            case {"op": "device_info", "connected": c,
                  "device_type": dt, "max_strength_a": ma, "max_strength_b": mb}:
                return CodecMessage(
                    op=OpCode.DEVICE_INFO, connected=c,
                    device_type=DeviceType(dt),
                    max_strength_a=ma, max_strength_b=mb,
                )
            case {"op": "ping", "t": t}:
                return CodecMessage(op=OpCode.PING, t=t)
            case {"op": "stop", "reason": r}:
                return CodecMessage(op=OpCode.STOP, reason=r)
            case _:
                raise ValueError(f"Unknown op: {data.get('op')}")

    @staticmethod
    def serialize(msg: CodecMessage) -> str:
        """将 ``CodecMessage`` 序列化回 JSON 字符串。"""
        data: dict[str, Any] = {"op": msg.op.value}
        for key in ("status", "data", "key", "value", "connected",
                     "device_type", "max_strength_a", "max_strength_b",
                     "reason", "t"):
            val = getattr(msg, key)
            if val is not None:
                data[key] = val.value if isinstance(val, StrEnum) else val
        return json.dumps(data)

File: sdk/typescript/src/codec.ts
/**
 * 消息类型与编解码器（序列化 / 反序列化）。
 * 与 Python SDK 的 `dghub_sdk.codec` 一一对应（协议完全一致）。
 */

import { Action, Channel, DeviceType, LogLevel, OpCode, StrengthMode } from "./enums.js";

/** `Codec.parse()` 的类型化结果。通过 `.op` 判断哪些字段有值。 */
export interface CodecMessage {
  op: OpCode;
  // hello_ack
  status?: string;
  // config
  data?: Record<string, unknown>;
  // config_changed
  key?: string;
  value?: boolean | number | string;
  // device_info
  connected?: boolean;
  deviceType?: DeviceType;
  maxStrengthA?: number;
  maxStrengthB?: number;
  // stop
  reason?: string;
  // ping
  t?: number;
}

/** trigger 消息参数（对应 Python `Codec.trigger` 的命名参数）。 */
export interface TriggerOptions {
  action?: Action;
  deltaPct?: number;
  strengthMode?: StrengthMode;
  durationS?: number;
  preset?: string;
  channel?: Channel;
  label?: string | null;
  username?: string | null;
  name?: string | null;
  cause?: string | null;
  pulseName?: string | null;
  targetId?: string | null;
}

/** event 消息参数（对应 Python `Codec.event` 的命名参数）。 */
export interface EventOptions {
  username?: string | null;
  strengthPct?: number | null;
  duration?: number;
  eventId?: string | null;
  cause?: string | null;
  pulseName?: string | null;
  fromPct?: number | null;
  toPct?: number | null;
  deltaPct?: number | null;
  targetId?: string | null;
}

function putIfNotNull(msg: Record<string, unknown>, key: string, value: unknown): void {
  if (value !== null && value !== undefined) {
    msg[key] = value;
  }
}

/** 所有消息编解码的命名空间类，无需实例化。 */
export class Codec {
  /** 构建 hello 握手消息。 */
  static hello(token: string, manifest: Record<string, unknown>): string {
    if (!token) {
      throw new Error("token is required");
    }
    if (typeof manifest !== "object" || manifest === null) {
      throw new TypeError("manifest must be an object");
    }
    return JSON.stringify({ op: "hello", token, manifest });
  }

  /** 构建统一触发消息。action 包含波形（BOTH / WAVEFORM）且 preset 为空时抛错。 */
  static trigger(options: TriggerOptions = {}): string {
    const action = options.action ?? Action.BOTH;
    if ((action === Action.BOTH || action === Action.WAVEFORM) && !options.preset) {
      throw new Error("preset is required when action includes waveform");
    }
    const msg: Record<string, unknown> = {
      op: "trigger",
      action: action,
      delta_pct: options.deltaPct ?? 0,
      strength_mode: options.strengthMode ?? StrengthMode.ROLLBACK,
      duration_s: options.durationS ?? 1.0,
      preset: options.preset ?? "",
      channel: options.channel ?? Channel.BOTH,
    };
    putIfNotNull(msg, "label", options.label);
    putIfNotNull(msg, "username", options.username);
    putIfNotNull(msg, "name", options.name);
    putIfNotNull(msg, "cause", options.cause);
    putIfNotNull(msg, "pulse_name", options.pulseName);
    putIfNotNull(msg, "target_id", options.targetId);
    return JSON.stringify(msg);
  }

  /** 构建一次性事件消息。 */
  static event(label: string, name: string, options: EventOptions = {}): string {
    if (!label) {
      throw new Error("label is required");
    }
    if (!name) {
      throw new Error("name is required");
    }
    const msg: Record<string, unknown> = {
      op: "event",
      label,
      name,
      duration: options.duration ?? 1.0,
    };
    putIfNotNull(msg, "username", options.username);
    putIfNotNull(msg, "strength_pct", options.strengthPct);
    putIfNotNull(msg, "event_id", options.eventId);
    putIfNotNull(msg, "cause", options.cause);
    putIfNotNull(msg, "pulse_name", options.pulseName);
    putIfNotNull(msg, "from_pct", options.fromPct);
    putIfNotNull(msg, "to_pct", options.toPct);
    putIfNotNull(msg, "delta_pct", options.deltaPct);
    putIfNotNull(msg, "target_id", options.targetId);
    return JSON.stringify(msg);
  }

  /** 构建仅波形的脉冲消息。 */
  static pulse(preset: string, channel: Channel = Channel.BOTH, targetId?: string | null): string {
    if (!preset) {
      throw new Error("preset is required");
    }
    const msg: Record<string, unknown> = { op: "pulse", preset, channel };
    putIfNotNull(msg, "target_id", targetId);
    return JSON.stringify(msg);
  }

  /** 构建 set_strength 消息。 */
  static setStrength(channel: Channel, pct: number, targetId?: string | null): string {
    if (!Number.isInteger(pct) || pct < 0 || pct > 100) {
      throw new Error("pct must be 0-100");
    }
    const msg: Record<string, unknown> = { op: "set_strength", channel, pct };
    putIfNotNull(msg, "target_id", targetId);
    return JSON.stringify(msg);
  }

  /** 构建 adjust_strength 消息。 */
  static adjustStrength(channel: Channel, deltaPct: number, targetId?: string | null): string {
    if (!Number.isInteger(deltaPct) || deltaPct < -100 || deltaPct > 100) {
      throw new Error("delta_pct must be -100 to 100");
    }
    const msg: Record<string, unknown> = { op: "adjust_strength", channel, delta_pct: deltaPct };
    putIfNotNull(msg, "target_id", targetId);
    return JSON.stringify(msg);
  }

  /** 构建状态上报消息。 */
  static status(fields: Record<string, unknown>): string {
    if (typeof fields !== "object" || fields === null) {
      throw new TypeError("fields must be an object");
    }
    return JSON.stringify({ op: "status", fields });
  }

  /** 构建日志消息。 */
  static log(level: LogLevel, message: string): string {
    return JSON.stringify({ op: "log", level, message });
  }

  /** 构建 set_config 消息，用于持久化运行时数据。 */
  static setConfig(key: string, value: unknown): string {
    return JSON.stringify({ op: "set_config", key, value });
  }

  /** 将 JSON 字符串解析为类型化的 CodecMessage。 */
  static parse(raw: string): CodecMessage {
    const data = JSON.parse(raw) as Record<string, unknown>;
    const op = data.op as string;
    switch (op) {
      case "hello_ack": {
        const fields: Record<string, unknown> = { ...data };
        delete fields.op;
        return { op: OpCode.HELLO_ACK, data: fields };
      }
      case "config":
        return { op: OpCode.CONFIG, data: (data.data ?? {}) as Record<string, unknown> };
      case "config_changed":
        return { op: OpCode.CONFIG_CHANGED, key: data.key as string, value: data.value as boolean | number | string };
      case "device_info":
        return {
          op: OpCode.DEVICE_INFO,
          connected: data.connected as boolean,
          deviceType: DeviceType[(data.device_type as string)?.toUpperCase() as keyof typeof DeviceType] ?? DeviceType.UNKNOWN,
          maxStrengthA: data.max_strength_a as number,
          maxStrengthB: data.max_strength_b as number,
        };
      case "ping":
        return { op: OpCode.PING, t: data.t as number };
      case "stop":
        return { op: OpCode.STOP, reason: data.reason as string };
      default:
        throw new Error(`Unknown op: ${op}`);
    }
  }

  /** 将 CodecMessage 序列化回 JSON 字符串。 */
  static serialize(msg: CodecMessage): string {
    const data: Record<string, unknown> = { op: msg.op };
    for (const key of [
      "status", "data", "key", "value", "connected",
      "deviceType", "maxStrengthA", "maxStrengthB", "reason", "t",
    ] as const) {
      const val = msg[key];
      if (val !== undefined && val !== null) {
        if (key === "deviceType") {
          data.device_type = (val as DeviceType) || DeviceType.UNKNOWN;
        } else if (key === "maxStrengthA") {
          data.max_strength_a = val;
        } else if (key === "maxStrengthB") {
          data.max_strength_b = val;
        } else {
          data[key] = val;
        }
      }
    }
    return JSON.stringify(data);
  }
}
